Initial WorkflowStepStore and Step and Task model.

srallen opened the pull request:
Package: lib-classifier

For #123.

Describe your changes:
This is an initial pass at adding a WorkflowStepStore, a Task model, and a Step model. I'd like an initial conceptual review on this before progressing much further. We should also discuss and approve #126 first before progression much more than this too. 

# Review Checklist

## General

- [ ] Are the tests passing locally and on Travis?
- [ ] Is the documentation up to date?
- [ ] Is the changelog updated?

## Apps

- [ ] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [ ] Can you `rm -rf node_modules/ && npm install` and app works as expected?

srallen commented on packages/lib-classifier/src/store/WorkflowStepStore.js:
```diff
@@ -0,0 +1,87 @@
+      if (taskKeys.length > 0) {
+        taskKeys.forEach((taskKey) => {
+          // Set tasks object as a MobX observable JS map in the store
+          self.tasks.set(taskKey, workflow.tasks[taskKey])
```
I might want to use `put` here instead of `set`, but the only place I see this described in in the MST docs about Map types: https://github.com/mobxjs/mobx-state-tree/blob/master/API.md#typesmap

I don't see it in the MobX docs itself...

srallen replied:
Looking over the source for MobX and MST, it appears to only be for MST.

srallen commented on packages/lib-classifier/src/store/Step.js:
```diff
@@ -0,0 +1,9 @@
+
+const Step = types.model('Step', {
+  next: types.optional(types.string, ''), // We have this optionally to support recursive workflows
+  stepKey: types.identifier,
```
This is duplicating the key of the Map key-pair, but the model needs an identifier too I think? I'm not sure. This example shows a separate identifier from the key of the key-value pair of the Map: https://github.com/mobxjs/mobx-state-tree/blob/master/API.md#typesmap

srallen replied:
I think I'm doing this right after a reread of the docs.

rogerhutchings commented on packages/lib-classifier/src/store/WorkflowStepStore.js:
```diff
@@ -0,0 +1,87 @@
+        const workflow = getRoot(self).workflows.active
+        if (workflow) {
+          self.reset()
+          if (workflow.steps.size > 0) {
```
I don't think there would be any steps on the workflow resource here - there would be tasks, but steps would be kept in this store.

Side thought: we could trash the existing workflow store and this one, and have a single store for the workflow which represents it in terms of steps and tasks.

srallen replied:
I think I'm still working through that conceptually and I just added the existential check just for safety when writing out the skeleton code. It really makes sense to me to have a separate WorkflowStep store, but could be convinced of a single store for the whole workflow resource.

rogerhutchings replied:
Let's leave them as separate stores for now. My feeling was that we don't need a store for the backend representation of the workflow - which doesn't get directly used - but that might change, especially while we're working this out

srallen replied:
I'm looking over this again and I'm confused about why the workflow resource here wouldn't have any steps. `active` references the `Workflow` model, which should be the workflow resource and I'm proposing the in updated ADR that we add a `steps` property onto the resource itself that is stored in Panoptes. Am I misunderstanding how references in MST work?

rogerhutchings replied:
Ahhhh - I forgot about the proposed resource change, apologies

srallen replied:
Haha no worries. I reread the code and then the docs and then the code again and confused myself before I realized that I think I'm on the right track here, but wanted to confirm with you. 😆

srallen commented on packages/lib-classifier/src/store/WorkflowStepStore.js:
```diff
@@ -0,0 +1,87 @@
+      if (snapshot.type === 'MultipleChoiceTask') return MultipleChoiceTask;
+    }, SingleChoiceTask, MultipleChoiceTask)))
+  })
+  .views(
```
Oops! I definitely missed that.

srallen commented on packages/lib-classifier/src/store/Task.js:
```diff
@@ -0,0 +1,122 @@
+import { types } from 'mobx-state-tree'
```
Yep, I was thinking that too, but just dumped everything in to a single file at first pass.

srallen commented on packages/lib-classifier/src/store/Step.js:
```diff
@@ -0,0 +1,9 @@
+import { types } from 'mobx-state-tree'
+
+const Step = types.model('Step', {
+  next: types.optional(types.string, ''), // We have this optionally to support recursive workflows
```
I'm not sure. On Panoptes, I'm proposing a string of the step key be saved here, however, we can make it a reference here in the store? I'm having a bit difficulty about what exactly the model should do. Should it be a representation of what we expect from Panoptes or of what we plan on storing for the client app... I'm already planning on conversion of the `tasks` object to an ES6 Map, so it might be ok if we document that what is stored in the resource is just a string of the step key.

srallen commented:
> I'm getting an MST error saying functions can't be used as properties. I'm having a further look...

I think the issue is that I was reading the docs for version 3 and we're still on version 2, so some of the types are throwing this error. I've resolved most of them, but I am having issues with the union type. I also have an experimental branch of updating to version 3 to see what's involved.

srallen commented:
I'm going to continue to work on this using MST v3 assuming that #136 will merge first.

srallen commented:
This is now rebased with the new models now correctly working. I've got the union model working now too. The docs aren't very clear, but if you're supplying a custom dispatcher function to tell it how to resolve, you have to pass it in as an object for the union function call, i.e. `types.union({ dispatcher: (snapshot) => Type })`

FWIW, the best way to think of this is that a compose type is like a merge whereas a union type is like a switch statement. Using union for the tasks makes sense to me here, since it'll be a map of a union. We don't have a consistent model for the tasks in the tasks map, so let's give it a switch statement of possibilities and then tell it how to resolve. (Calling this a union doesn't initially make sense either, but I think MST is borrowing heavily from typescript, so maybe a basic overview of that might help here).

srallen commented:
@rogerhutchings Could I get another review of this? I'm likely to branch off this to start the classification store.

srallen commented:
I'm not seeing where this is going wrong, and it's also true for the new ClassificationStore I'm working on too. @rogerhutchings any suggestions?

rogerhutchings commented:
Ah, now I think I have it - the steps store isn't being observed by anything, so never gets attached. I tried throwing steps into a `storeMapper` function and my `console.log`s started appearing.

srallen commented:
I'm not sure if that's it because I still can't get a log to happen in the WorkflowStepStore's `afterAttach` event and I tried making the `TaskArea` component an observer. Could you share more details about what you did?

srallen commented:
Ok, ready for this to be reviewed again! A few more changes include:
- Renaming `ClassifierStore` to `SubjectViewerStore` to make it clear what state that this store is for. I found it confusing to have this store named this when it is only the states for the subject viewer pan/zoom tools.
- Rename the top level store key name for `steps` to `workflowSteps`. `steps.steps` is not good. 
- Add an observer for the `workflowSteps` to `TaskArea`. This will likely change, but gets the store to attach!
- Added some backwards compatibility for the KittehZoo project's workflow so we can continue developing with it. If a workflow doesn't have defined steps, just store the tasks.
- Added a `DrawingTask` model because KittehZoo has a drawing task and if there isn't a defined model for a task type, MST will yell at you.

srallen commented:
We could think about writing a function that converts the tasks into steps. Unless it's a combo task or there's an added shortcut task, we can assume one task per step. It's an interesting idea that could help in the migration process for projects later.... but one to think about in a separate PR for sure.
